### PR TITLE
validate: verify possibly dangerous operations in CWL workflows

### LINKED ISSUE #458: validate: verify possibly dangerous operations
Building up on previous tasks of verifying environment images, see #457 for example, we can take validation further to perform some sanity checks on the workflow commands.

For example, it happened that a workflow was running `sudo /some/command/...` all the while asking to run the job as some kubernetes_uid. We should warn the user about these cases.

For example, the `validate --steps` could run a quick `grep` looking for `sudo`, and report findings.

Another example, operations like `cd /some/path` may be dangerous, if there is a future cp/wget/xrdcp` populating some internal directory hard-coded inside the container image, and not populating the usual workspace directory, with input files. We could do a quick grep helping to spot these situations.

IOW, we are not looking for the perfect solution, but a good first approach to catching these early on the reana-client side.

(E.g. the writing-to-local-ephemeral-storage situation will be best caught by the REANA platform on the server side, but that will become another issue.)

===

**Current status:**

- [x] Serial - https://github.com/reanahub/reana-client/pull/471
- [x] Yadage - https://github.com/reanahub/reana-client/pull/470/commits/9c886a1c11cc3524a2c9525e8a8850f2736e5aaf
- [x] CWL - https://github.com/reanahub/reana-client/pull/481

### PR DESCRIPTION
**Pull request by ParthS007**
closes #458

### CODE REVIEW

File: reana_client/validation.py [hunk 1]
```diff
@@ -521,6 +521,39 @@ def _validate_cwl_parameters(reana_yaml):
         )
         sys.exit(1)
 
+    dang_params = _validate_cwl_dangerous_operations(reana_yaml)
```
Comment (mvidalgarcia) [inline reana_client/validation.py:524]: We already have [`_validate_dangerous_operations`](https://github.com/reanahub/reana-client/blob/20dc10301b74766b3f9c27daba6608d1e65a3840/reana_client/validation.py#L640-L654), please use it to have consistent warning messages for all the workflow engines.

I see that for CWL the potential dangerous operations can be both in `baseCommand` and `arguments`. Some examples:
- https://github.com/reanahub/reana-demo-cdci-crab-pulsar-integral-verification/blob/2674d484cd2cb1ac8c00068e6bc53804be31b999/workflow/cwl/crab.cwl#L3-L7
- https://github.com/reanahub/reana-demo-root6-roofit/blob/a43db4aad1c3e1e2c3f7eb8d352169c0eba1beb6/workflow/cwl/fitdata.cwl#L20-L26

I wonder if we should also check `inputs` as they may contain some bash statements inside.. 🤔, although this wouldn't be a good practice at all so we can pass, it's users' resposibility.

So the idea would be simply passing the value of those fields to `_validate_dangerous_operations`, which will take care of doing the actual checks.

File: tests/test_validate_parameters.py [hunk 2]
```diff
@@ -61,6 +62,15 @@ def get_loaded_yaml(step_spec, input_spec):
...
 
+    # Dangerous operation used in parameter
+    with runner.isolated_filesystem():
+        reana_yaml = get_loaded_yaml(
+            cwl_workflow_spec_step, cwl_workflow_spec_dangerous_input_param
+        )
+        _validate_cwl_parameters(reana_yaml)
+        captured = capsys.readouterr()
+        assert 'Operation "sudo python" found in "baseCommand"' in captured.out
+
```
Comment (mvidalgarcia) [inline tests/test_validate_parameters.py:73]: We probably don't need tests for this. We're already testing the functionality [here](https://github.com/reanahub/reana-client/blob/20dc10301b74766b3f9c27daba6608d1e65a3840/tests/test_validate_parameters.py#L243).

File: reana_client/validation.py (hunk 1 above)
Comment (ParthS007) [reply reana_client/validation.py]: I am now using the `_validate_dangerous_operations` and warning are shown like as below. 

Screenshot 2021-03-18 at 12 14 50 PM

Comment (mvidalgarcia) [reply reana_client/validation.py]: Judging from your screenshot. `arguments` and `baseCommand` are not steps, steps are e.g. `gendata` and `fitdata` from `r-d-root6-roofit` example.

Comment (ParthS007) [reply reana_client/validation.py]: okay, Should I name them as `param` or `Operation "cd /" found in "arguments" might be dangerous` is enough?

Comment (mvidalgarcia) [reply reana_client/validation.py]: No, it's not enough. The idea here is that if the user has a workflow with tens of steps, is able to quickly identify where this dangerous operation is. Please check the output when introducing a dangerous operation for Serial or Yadage workflows in `r-d-root6-roofit`:

```console
$ reana-client validate -f reana-yadage.yaml               
==> Verifing REANA specification file... /Users/marco/code/reanahub/reana-demo-root6-roofit/reana-yadage.yaml
  -> SUCCESS: Valid REANA specification file.
==> Verifing workflow parameters and commands... 
  -> WARNING: Operation "cd /" found in step "gendata" might be dangerous.
  -> WARNING: Operation "sudo" found in step "fitdata" might be dangerous.
```

The output should be exactly the same for CWL.

Take into account that a step might have no name, (e.g. `r-d-worldpopulation`). In that case, we show the number.

```console
$ reana-client validate                     
==> Verifing REANA specification file... /Users/marco/code/reanahub/reana-demo-worldpopulation/reana.yaml
  -> SUCCESS: Valid REANA specification file.
==> Verifing workflow parameters and commands... 
  -> WARNING: Operation "sudo" found in step "0" might be dangerous.
```

Comment (ParthS007) [reply reana_client/validation.py]: The difference in CWL workflows is these `arguments` and `baseCommand` may/may not be be included in the steps. In `reana-demo-cdci-crab-pulsar-integral-verification` it is present in `workflow spec` directly not in any step.

Comment (mvidalgarcia) [reply reana_client/validation.py]: I see.. it seems that the cdci example is a bit special as it doesn't have a `Worlkflow` file but only a `CommandLineTool` one, which by checking the docs:

```
A CommandLineTool is a process characterized by the execution of a standalone, non-interactive program which is invoked on some input, produces output, and then terminates.
```

So it looks like a basic execution, no steps involved, which seems a bit weird to me.

So what we can do is making `step` optional in `_validate_dangerous_operations`, `None` by default. If `None`, we can omit the "found in step {}" part. So we can check if the workflow has steps, and pass the proper values. Please a a new test case for this in [test_validate_dangerous_operations](https://github.com/reanahub/reana-client/blob/20dc10301b74766b3f9c27daba6608d1e65a3840/tests/test_validate_parameters.py#L243).

Comment (ParthS007) [reply reana_client/validation.py]: As we see that this behavior is limited to only `cwl` workflow. what about sending `step=None` for CWL and keep as it is for other?

Comment (ParthS007) [reply reana_client/validation.py]: @mvidalgarcia  
Screenshot 2021-03-18 at 3 26 08 PM
How does it look now?

File: reana_client/validation.py [hunk 3]
```diff
@@ -637,21 +656,33 @@ def parse_params(stages):
...
-            display_message(
-                'Operation "{}" found in step "{}" might be dangerous.'.format(
-                    operation.strip(), step
-                ),
-                msg_type="warning",
-                indented=True,
-            )
+        if isinstance(command, str) and operation in command:
+            _display_message(operation, step, step_flag)
+        if isinstance(command, list):
```
Comment (mvidalgarcia) [inline reana_client/validation.py:680]: A command should never be a list, could we move this logic to the `_check_dangerous_operations` inside CWL and leave this function more generic?

File: reana_client/validation.py (hunk 1 above)
Comment (mvidalgarcia) [reply reana_client/validation.py]: > As we see that this behavior is limited to only `cwl` workflow. what about sending `step_flag=None` for CWL and keep as it is for other?

I don't see the benefit of adding a flag. The step itself can serve as a flag. If it's passed it means we have to display it on the message, if it's not passed, we don't show it.

File: reana_client/validation.py [hunk 4]
```diff
@@ -521,6 +521,23 @@ def _validate_cwl_parameters(reana_yaml):
         )
         sys.exit(1)
 
+    def _check_dangerous_operations(workflow):
+        """Check for dang_keys in workflow and validate operations."""
+        [
+            _validate_dangerous_operations(v, k, step_flag=None)
+            for k, v in workflow.items()
+            if k in dang_keys
+        ]
```
[the remark this answers is not in the document]
Comment (ParthS007) [reply reana_client/validation.py:530]: Yes we are not returning anything. Initially, I added just plain for loop and if condition then I changed like this. Now I think you are right, I am not using the list anywhere so better keep it plain with list comprehension.

File: reana_client/validation.py [hunk 5]
```diff
@@ -521,6 +521,29 @@ def _validate_cwl_parameters(reana_yaml):
...
+    def _check_dangerous_operations(workflow):
+        """Check for "baseCommand" and "arguments" in workflow.
+
+        If these keys are found, validate if they have dangerous operations.
+        """
+        for k, v in workflow.items():
+            if k in dang_keys:
+                if isinstance(v, list):
+                    for val in v:
+                        _validate_dangerous_operations(val, None)
```
Comment (mvidalgarcia) [inline reana_client/validation.py:533]: The idea here is to pass the step if it exists. Maybe we can pass `workflow.get('id')`, at least we would have some useful info for the user.

Comment (ParthS007) [reply reana_client/validation.py:533]: What would be the name of the `id` in the msg. `step`/`id` or something else?

Comment (ParthS007) [reply reana_client/validation.py:533]: If we pass the `id` from here, then we need the flag in `_validate_dangerous_operations` to check if it's coming from cwl to show the output msg accordingly.

Comment (mvidalgarcia) [reply reana_client/validation.py]: We can just treat it as the step, in the end is kind of similar.

File: reana_client/validation.py [hunk 6]
```diff
@@ -521,6 +521,31 @@ def _validate_cwl_parameters(reana_yaml):
         )
         sys.exit(1)
 
+    def _check_dangerous_operations(workflow):
+        """Check for "baseCommand" and "arguments" in workflow.
+
+        If these keys are found, validate if they have dangerous operations.
+        """
+        dang_keys = ["baseCommand", "arguments"]
+        for k in workflow.keys():
```
Comment (mvidalgarcia) [inline reana_client/validation.py:530]: I meant we don't need to iterate over all keys, as we're only interested in `baseCommand` and `arguments`. Something like this:

```python
cmd_keys = ["baseCommand", "arguments"]
for cmd_key in cmd_keys:
    if cmd_key in workflow:
        cmd_values = (
            workflow[cmd_key]
            if isinstance(workflow[cmd_key], list)
            else [workflow[cmd_key]]
        )
        for cmd_value in cmd_values:
            _validate_dangerous_operations(str(cmd_value), step=workflow.get("id"))
```

**Review by mvidalgarcia — APPROVED** [review]
Looking good 👍

**Pull request merged**